### ISSUE TITLE
Eventual memory exhaustion (OutOfMemoryError) with no apparent leak

### ISSUE THREAD

**Issue by gueugaie** [bytedeco/javacpp#239]
I've built a PDF processing pipeline, that ends up processing rasterized images using JavaCV, javaCPP / presets version is latest (1.4.1).

Context
----------

I've got to a point where I understood, at least, the most importants aspects of JavaCPP's Pointers and subtelties, which are not that trivial to get coming from a pure Java background.

E.g. i would do :

```
public ByteBuffer toJPG(Mat myImage) {
  BytePointer bp = new BytePointer();
  imencode(myImage, "jpg", bp)
  return bp.asBuffer();
}
```

Only to get a SIGSEGV at a later point when accessing the returned `ByteBuffer`. There are other variants of this "trick", but that is not the point here (just wanted to tell, some things may be obvious to those used to calling native code, and I'm not one of those).

The issue
------------

What bothers me, though, is that my service always ends up blowing up OutOfMemoryErrors (launched by the Pointer's dealloc method in its gc()/trim() loop). It takes time, but it does happen.

My allocation rate being pretty fast, although my service is launched with `-Xmx368m`, I've put `-Dorg.bytedeco.javacpp.maxphysicalbytes=2g`. Raising the `maxRetries` does not help, as I'll illustrate. The program starts using ~ 800Mb RSS, climbs to 2G, drops down, reclimbs, ... but will converge to 2GB and end up in the inability to alloc any new `Pointer`.

So I've set up a test loop using. I launch my pipeline as a REST service, and loop over a set of 300 files. Kind of `find . -print 0 | xargs -0 -I _FILE -P 1 curl _FILE`, although the first loop is 1 file at a time, then 5, then 1, ... To simulate various loads.

I run the service on a Ubuntu 16.04, Java 8u121, x86_64. I call `top` every second to graph the process's RSS, and this is what I get (graphing lowest RSS value of the last 10 seconds) :

![sans titre](https://user-images.githubusercontent.com/15906268/40651979-4f45e3c6-6337-11e8-9423-d35c6febef41.png)
 
The horizontal scale is in seconds. You see it slowly incresases. Trust me it never goes down, it eventually refuses to allocate any new Pointer, as the trim() process ends up never freeing any more bytes, the RSS stays flat at ~2GB.

I've dumped the java heap and confirmed NO instance of org.byteco is live. No DeallocatorReference is live either. The PhantomReference queue is empty.

The way I see it, either I'm doing something wrong, or JavaCV / openCV leaks memory, or native memory allocation suffers from fragmentation, which, makes it impossible to release part of it back to the OS (e.g. it can not diminish its RSS), which locks javaCPP's allocation process.

So...

What I've tried (part 1)
-----------------------------
Configuring `-Dorg.bytedeco.javacpp.maxphysicalbytes=0`. 'Solves' the issue, in that OutOfMemory does no occur, but the process eats up all available memory => not an option

What I've tried (part 2)
-----------------------------
Tracing memory leaks.
I already had Java heap dumps => nothing leaks from them.
So switched glibc's malloc for [jemalloc](https://github.com/jemalloc/jemalloc/wiki/Use-Case%3A-Leak-Checking). 

I ran 2 loops of the process, and here is the result :
[jeprof.pdf, live allocated blocks when the process has exited from main](https://github.com/bytedeco/javacpp/files/2047730/jeprof.pdf)
From what I can tell, no javaCV / openCV object stays live in the native heap (I did some poking aroung, you can't miss them when there are, you have a lot of cv::fastAlloc blocks all aroung). Here, only GC and JVM related stuff.

So : no memory leaks, either Java or Native.

![usage_jemalloc_leak](https://user-images.githubusercontent.com/15906268/40652563-d462b6fa-6338-11e8-8065-5fa5eb2a9574.png)

Still, the RSS rises, and rises, and rises... More slowly though.

What I've tried (part 3)
-----------------------------
 So I launched the same test, again, without leak checking, just switching glibc's malloc to jemalloc and setting both Xmx and Xms on the JVM (on a hunch : if you tell the JVM to allocate heap at once, then you have a lower chance of it fragmenting the native memory). Ran it for *much* longer.

![mem-min](https://user-images.githubusercontent.com/15906268/40652641-14b972e8-6339-11e8-9c46-e4a5a85c3be7.png)

Yes the RSS still rises. But far more slowly than it did in the first test (here, you have 7 hours worth of usage, and the RSS can still drop down to ~700MB, while glibc's sample at the top could not drop down 1.5GB after only 2 hours).

So that's cool, but I may not have the luxury of switching malloc implementations on a production server...

Questions
------------

Am I doing something wrong ? What can I try ? 

I have a feeling the `maxphysicalbytes` stuff is a somehow wrong way of thinking about memory usage. 
If you want to limit native memory allocation, but you only check the RSS size for that (without checking how much of that size is actually free memory) then you're basically saying "the restaurant is fully booked tonight" just because there is one single person at the furthest table.

Fragmentation seems to be the key... ?

So maybe I could ulimit the Java process to 2GB and set `maxphysicalbytes` to zero... Would that be worth a test ?

Maybe the `gc()/trim()` loop has an issue, in that it tries to `trim()` the RSS *after* it has allocated the current pointer, which may live at the top of the heap, preventing `trim()` to actually trim anything (e.g. one should gc/trim either before alloc() of after dealloc() but not in-between )?

Do you have any suggestion ?

**Comment by saudet** [comment]
`Pointer.asBuffer()` doesn't copy the underlying data, it is owned by the `BytePointer` in the case above, while there is no way to make `Buffer` keep a reference to `Pointer`, so when such a `Pointer` goes out of scope and gets garbage collected, the underlying memory vanishes and when accessed with the wrapped `Buffer` can result in a segmentation fault. If you need to have a standalone `Buffer`, you'll need to copy the data out of the `Pointer`.

Yes, slowly rising RSS is a sign of memory fragmentation. You'll need to rearrange how your program makes allocations. You could for example allocate a huge `ByteBuffer`, pass that to `imencode()`, and cache it for a later call. That way you avoid both memory fragmentation caused by that allocation and since the `Buffer` owns the memory, we don't need to keep a reference to a `Pointer` somewhere.

BTW, jemalloc is more resistant to memory fragmentation than glibc, so if you link your libraries with it instead of glibc, that *should* help as well.

**Comment by gueugaie** [comment]
Thanks for commenting. I had figured it out by myself about the Pointer / Buffer needing to be both retained.

I get how I could "pool" `ByteBuffer`s and reuse them, at least conceptually, for the image encoding processes. But most of my allocation is actually done through `Mat` instances, where I don't really see how I could do the same (I'd have to make all my images the same width/height/depth to prevent openCV from reallocating the underlying storage, I guess... that's not a practical use of the API at all). Do you have any thought here ?

**Comment by saudet** [comment]
If most of these allocations are only temporary, deallocating them immediately as soon as we don't need them helps a lot too. This is much easier to do now with PointerScope, see commit https://github.com/bytedeco/javacpp/commit/d98d323778c4b73865d19c7fc15c8380b130f5e0.

**Comment by gueugaie** [comment]
Thanks. I spent a few hours wrapping "everything" in try/finally, and it does improve things.

I also played with `-nopointergc`, mostly to know if I had caught everything. I noticed that when the mode is activated, `NativeDeallocatorReferences` do pile up in the `next`/`previous` linked list, and nothing ever clears them up, eventually leading to a Java side OOM. (Then again, maybe I'm doing something wrong).

Reproduction is easy as can be : 

```
    public static void main(String[] args) throws Exception {
        int blackHole = 0;
        for (int i = 0; i < 100_000; i++) {
            try (opencv_core.Mat testObject = new opencv_core.Mat(10, 10, CV_8UC1)) {
                blackHole += testObject.type();
            }
        }
        Class<?> deallocClass = Class.forName(Pointer.class.getCanonicalName() + "$DeallocatorReference");
        Field headField = deallocClass.getDeclaredField("head");
        headField.setAccessible(true);
        Field nextField = deallocClass.getDeclaredField("next");
        nextField.setAccessible(true);
        int nbOfDeallocators = 0;
        synchronized (deallocClass) {
            Object deallocRef = headField.get(null);
            do {
                deallocRef = nextField.get(deallocRef);
                nbOfDeallocators++;
            } while (deallocRef != null);

        }
        System.out.println(nbOfDeallocators);
        System.out.println(blackHole);
    }
```

You'll see 100 000 elements (actually, NativeDeallocators) are held by `DeallocatorReference`'s static linked chain, and have no chance of ever getting collected.

**Comment by saudet** [comment]
Oops, sorry about that. I've fixed this issue in commit https://github.com/bytedeco/javacpp/commit/7a90061d996e80197274564b573b92c39a51e311.

Note that this mode shouldn't be used in production and is only really useful for debugging purposes.

**Comment by gueugaie** [comment]
Thanks a lot. I'll follow up if I can reach a stable state (e.g. tens of days under 2GB), for new and using glibc's malloc, I'm still far from it.
I will also dig further on the pure openCV implementations to try and see what can be done there. Feel free to close the thread.

**Comment by saudet** [comment]
The fix is now included in version 1.4.2, thanks for reporting!

As for memory fragmentation, using `PointerScope` should prevent any issues related to that:
http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/
Let me know if this is still a problem though or have other ideas that we could integrate to JavaCPP. Thanks for your time on this!

**Comment by xuhao417347761** [comment]
> Maybe the `gc()/trim()` loop has an issue, in that it tries to `trim()` the RSS _after_ it has allocated the current pointer, which may live at the top of the heap, preventing `trim()` to actually trim anything (e.g. one should gc/trim either before alloc() of after dealloc() but not in-between )?

Hi, gueugaie

     Recently I am working on building an automatic java memory leak detector. It seems this is a good candidate case to examine the memory leak detector. Even though this bug is fixed, could you help me on reproduce this bug?

Yours
Hao
